### PR TITLE
added support for custom namespace

### PR DESCRIPTION
**Pull request by EnoahNetzach**
@rande solved issue https://github.com/sonata-project/SonataEasyExtendsBundle/issues/60, adding support for a custom namespace via the command line option `--namespace`, where a placeholder could be used (`:vendor`) for substituting the extended bundle's vendor name.
This defaults to `Application\:vendor`, as before.

Also mustache is now used for substituting `{{ namespace }}` in the ord, odm and phpcr mappings.
Other Sonata bundles should be uploaded (now it's hardcoded `Application\Sonata`).

### CODE REVIEW

**Comment by 0xMatt** [comment]
:+1: this should have always been the case

**Comment by MattWohler** [comment]
:+1: no brainer

**Comment by core23** [comment]
@EnoahNetzach Please rebase

File: Generator/OrmGenerator.php [hunk 1]
```diff
@@ -55,7 +55,14 @@ public function generateMappingEntityFiles(OutputInterface $output, BundleMetada
                 $output->writeln(sprintf('   ~ <info>%s</info>', $fileName));
             } else {
                 $output->writeln(sprintf('   + <info>%s</info>', $fileName));
-                copy($src_file, $dest_file);
+
+                $mappingEntityTemplate = file_get_contents($src_file);
+
+                $string = Mustache::replace($mappingEntityTemplate, array(
+                    'namespace'    => $bundleMetadata->getExtendedNamespace()
```
Comment (OskarStark) [inline Generator/OrmGenerator.php]: please remove the whitespace before `=>`

Comment (core23) [inline Generator/OrmGenerator.php]: Please fix and rebase @EnoahNetzach

**Comment by EnoahNetzach** [comment]
@core23 sorry for the delay

File: Resources/doc/reference/why.rst [hunk 2]
```diff
@@ -104,3 +104,5 @@ At last you can run:
 .. note::
 
     Note that the `--dest` option allows you to choose the target directory, such as `src`. Default destination is `app/`.
+
+    Also note that the `--namespace` option allows you to choose the base namespace, such as `Application\\Sonata`. Default destination is `Application\\:vendor`.
```
Comment (OskarStark) [inline Resources/doc/reference/why.rst]: instead of `Alslo note...`, please add a second `.. note::`-Block

**Comment by OskarStark** [comment]
except my small comment this PR looks good to me!

@core23 why is travis not green, didn't you fixed thes travis config?

@greg0ire can you have a second look?

Thanks

**Comment by greg0ire** [comment]
Please keep your commit message subject length under 50 chars.

Please read this :

- http://stopwritingramblingcommitmessages.com/
- http://rakeroutes.com/blog/deliberate-git/

File: Command/GenerateCommand.php [hunk 3]
```diff
@@ -63,8 +64,21 @@ protected function execute(InputInterface $input, OutputInterface $output)
             $dest = $this->getContainer()->get('kernel')->getRootDir();
         }
 
+        $namespaceOption = $input->getOption('namespace');
+        if ($namespaceOption) {
+            $namespace = $namespaceOption;
+            if (!preg_match('/^(?:(?:[[:alnum:]]+|:vendor)\\?)+$/', $namespace)) {
+                $output->writeln('');
+                $output->writeln(sprintf('<error>The provided namespace \'%s\' is not a valid namespace!</error>', $namespaceOption));
```
Comment (greg0ire) [inline Command/GenerateCommand.php]: Please linebreak this.

File: Command/GenerateCommand.php [hunk 4]
```diff
@@ -63,8 +64,21 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+                return 0;
+            }
+        } else {
+            $namespace = 'Application\:vendor';
+        }
+
         $configuration = array(
-            'application_dir' =>  sprintf("%s/Application", $dest)
+            'application_dir' =>  sprintf("%s%s%s", $dest, DIRECTORY_SEPARATOR, str_replace('\\', DIRECTORY_SEPARATOR, $namespace)),
```
Comment (greg0ire) [inline Command/GenerateCommand.php]: Please linebreak this.

File: Generator/BundleGenerator.php [hunk 5]
```diff
@@ -67,7 +67,8 @@ protected function generateBundleDirectory(OutputInterface $output, BundleMetada
      */
     protected function generateBundleFile(OutputInterface $output, BundleMetadata $bundleMetadata)
     {
-        $file = sprintf('%s/Application%s.php', $bundleMetadata->getExtendedDirectory(), $bundleMetadata->getName());
+        $application = explode('\\', $bundleMetadata->getExtendedNamespace())[0];
+        $file = sprintf('%s/%s%s.php', $bundleMetadata->getExtendedDirectory(), $application, $bundleMetadata->getName());
```
Comment (greg0ire) [inline Generator/BundleGenerator.php]: Please linebreak this

File: Resources/doc/reference/introduction.rst [hunk 6]
```diff
@@ -16,3 +16,7 @@ The command line will generate:
 
 You can optionally define a ``--dest`` option to the command with the target directory for the extended bundle creation.
 By default, this is set to ``app`` but you should probably set it to ``src``.
+
+You can optionally define also a ``--namespace`` option to the command with the namespace for the extended bundle classes and directory structure.
```
Comment (greg0ire) [inline Resources/doc/reference/introduction.rst]: Please linebreak this (`gqq` in vim).

File: Tests/Bundle/BundleMetadataTest.php [hunk 7]
```diff
@@ -65,7 +75,7 @@ public function testBundleName()
     {
         $bundle = new \Sonata\AcmeBundle\AcmeBundle();
 
-        $bundleMetadata = new BundleMetadata($bundle, array('application_dir' => 'Application'));
+        $bundleMetadata = new BundleMetadata($bundle, array('application_dir' => 'Application', 'namespace' => 'Application'));
```
Comment (greg0ire) [inline Tests/Bundle/BundleMetadataTest.php]: Please keep all these lines under 120 chars.

File: Command/GenerateCommand.php [hunk 8]
```diff
@@ -67,7 +67,7 @@ protected function execute(InputInterface $input, OutputInterface $output)
         $namespaceOption = $input->getOption('namespace');
         if ($namespaceOption) {
             $namespace = $namespaceOption;
-            if (!preg_match('/^(?:(?:[[:alnum:]]+|:vendor)\\?)+$/', $namespace)) {
+            if (!preg_match('/^(?:(?:[[:alnum:]]+|:vendor)\\\\?)+$/', $namespace)) {
```
Comment (greg0ire) [inline Command/GenerateCommand.php]: Your commit message should say what you are doing and why, so

```
detect one more slash with regex

No idea why you do that, please replace with explanation.
```

File: Command/GenerateCommand.php [hunk 9]
```diff
@@ -44,7 +44,7 @@ protected function configure()
 
         $this->addArgument('bundle', InputArgument::IS_ARRAY, 'The bundle name to "easy-extends"');
         $this->addOption('dest', 'd', InputOption::VALUE_OPTIONAL, 'The base folder where the Application will be created', false);
-        $this->addOption('namespace', 'n', InputOption::VALUE_OPTIONAL, 'The namespace for the classes', false);
+        $this->addOption('namespace', 'm', InputOption::VALUE_OPTIONAL, 'The namespace for the classes', false);
```
Comment (greg0ire) [inline Command/GenerateCommand.php]: Why ? How is this a fix?

File: Generator/PHPCRGenerator.php [hunk 10]
```diff
@@ -55,7 +55,14 @@ public function generateMappingDocumentFiles(OutputInterface $output, BundleMeta
...
                 $output->writeln(sprintf('   + <info>%s</info>', $fileName));
-                copy($src_file, $dest_file);
+
+                $mappingEntityTemplate = file_get_contents($src_file);
+
+                $string = Mustache::replace($mappingEntityTemplate, array(
+                    'namespace'    => $bundleMetadata->getExtendedNamespace()
+                ));
+
+                file_put_contents($dest_file, $string);
```
Comment (greg0ire) [inline Generator/PHPCRGenerator.php]: Please explain why you are doing all this in your commit message.

**Comment by greg0ire** [comment]
Please rebase to make the merge commit disappear, and please fixup the second "whitespace removal" commit into the first one.

**Comment by Soullivaneuh** [comment]
According to the new [Sonata version management](https://github.com/sonata-project/SonataAdminBundle/issues/3053) and next major [release plan](https://github.com/sonata-project/SonataAdminBundle/issues/3731), this project has been refactored regarding branching and versioning.

If you see this message, your PR concerns a patch or a minor release and is not targeting the right branch.

So I'm closing this one, but don't see it as a refusal. If you think your work is still relevant and want to continue, feel free to reopen it on the right branch (e.g. the default one).

Regards.

**Pull request closed without merging**